src/powerquery-language-services/analysis.ts

```diff
@@ -153,24 +137,72 @@ abstract class AnalysisBase implements Analysis {
             return LanguageServiceUtils.EmptySignatureHelp;
         }
 
-        // TODO: add tracing/logging to the catch()
-        const librarySignatureHelp: Promise<SignatureHelp | null> = this.librarySymbolProvider
-            .getSignatureHelp(context)
-            .catch(() => {
-                // tslint:disable-next-line: no-null-keyword
-                return null;
-            });
-
-        const [libraryResponse] = await Promise.all([librarySignatureHelp]);
-
-        return libraryResponse ?? LanguageServiceUtils.EmptySignatureHelp;
+        // Result priority is based on the order of the symbol providers
+        return AnalysisBase.resolveProviders(
+            AnalysisBase.createSignatureHelpCalls(context, [
+                this.localSymbolProvider,
+                this.environmentSymbolProvider,
+                this.librarySymbolProvider,
+            ]),
+            LanguageServiceUtils.EmptySignatureHelp,
+        );
     }
 
     public abstract dispose(): void;
 
     protected abstract getLexerState(): WorkspaceCache.LexerCacheItem;
     protected abstract getText(range?: Range): string;
 
+    private static createCompletionItemCalls(
+        context: CompletionItemProviderContext,
+        providers: CompletionItemProvider[],
+    ): Promise<CompletionItem[]>[] {
+        // TODO: add tracing to the catch case
```

Comment:
Maybe "handle" the catch case of each of these by creating an empty/no-op `Trace.error(...)` method. I think it'd be easier to implement a single function later on than making sure we add tracing to each spot.

Reply:
You're welcome to take a stab a trace interface if you're feeling up to it. I was putting this one off as well, since we need something that will work in the VS Code / Language Server environment, as well as all Power Query hosts. I was going to wait until we're a bit more integrated into PQ before looking to see what makes sense there.